types/aws-lambda/aws-lambda-tests.ts

```diff
@@ -812,6 +849,34 @@ apiGtwProxyHandler = proxyHandler;
 const codePipelineHandler: AWSLambda.CodePipelineHandler = (event: AWSLambda.CodePipelineEvent, context: AWSLambda.Context, cb: AWSLambda.Callback<void>) => {};
 
 const cloudFrontRequestHandler: AWSLambda.CloudFrontRequestHandler = (event: AWSLambda.CloudFrontRequestEvent, context: AWSLambda.Context, cb: AWSLambda.CloudFrontRequestCallback) => {
+    event = CloudFrontRequestWithCustomOriginEvent;
+    // $ExpectType CloudFrontRequestEvent
+    event;
+    let request = event.Records[0].cf.request;
+    if (request.origin && request.origin.custom) {
+        request.origin.custom.domainName;
+        request.origin.custom.domainName = "example2.com";
+    }
+
+    if (request.origin) {
+      // $ExpectError
+      request.origin.s3;
+    }
```

Comment:
Not sure what this and `.custom` below is checking for? If these were testing that the `.s3` was not accessible inside the `if (request.origin.custom)` due to flow analysis, that would make sense, but at the moment this seems to be some other error.

Reply:
Ahh, it's failing the tests, which is what I would be expecting. Try moving these inside the `if`s above each: e.g. `.s3` is not accessible if `.custom` is, and vice versa.

Reply:
Not sure if this is what you meant:
```
    let request = event.Records[0].cf.request;
    if (request.origin && request.origin.custom) {
        request.origin.custom.domainName;
        request.origin.custom.domainName = "example2.com";

        // $ExpectError
        request.origin.s3;
    }
```
Apparently this found no error.

Reply:
Try reading or writing to it.

Reply:
OK, maybe that was my misunderstanding of how these tests are supposed to work. Looking at them I assumed that simply referencing an object was enough to assert something.

Anyway:
```
    let request = event.Records[0].cf.request;
    if (request.origin && request.origin.custom) {
        request.origin.custom.domainName;
        request.origin.custom.domainName = "example2.com";

        // $ExpectError
        const s3 = request.origin.s3;

        // $ExpectError
        request.origin.s3 = s3;
    }
```
Sadly neither of these last two assignments fail.

Reply:
Sorry, I wasn't too clear, [this playground link](http://www.typescriptlang.org/play/#src=type%20Y%20%3D%20%7B%20c%3A%20number%20%7D%3B%0D%0Atype%20X%20%3D%0D%0A%20%20%20%20%7C%20%7B%20a%3A%20Y%2C%20b%3F%3A%20never%20%7D%0D%0A%20%20%20%20%7C%20%7B%20a%3F%3A%20never%2C%20b%3A%20Y%20%7D%3B%0D%0A%0D%0Aconst%20x1%3A%20X%20%3D%20%7B%20a%3A%20123%20%7D%3B%0D%0Aconst%20x2%3A%20X%20%3D%20%7B%20b%3A%20456%20%7D%3B%0D%0Alet%20y%3A%20Y%20%3D%20%7B%20c%3A%20789%20%7D%3B%0D%0A%0D%0Aif%20(x1.a)%20%7B%0D%0A%20%20%20%20x1.a%20%3D%20y%3B%20%2F%2F%20OK!%0D%0A%20%20%20%20y%20%3D%20x1.a%3B%20%2F%2F%20OK!%0D%0A%0D%0A%20%20%20%20x1.b%3B%20%2F%2F%20no%20error!%20reads%20undefined%0D%0A%20%20%20%20x1.b%20%3D%20y%3B%20%2F%2F%20error%2C%20assign%20Y%20to%20undefined%0D%0A%20%20%20%20y%20%3D%20x1.b%3B%20%2F%2F%20error%2C%20assign%20undefined%20to%20Y%0D%0A%7D) shows what I mean.

Basically `$ExpectError` just makes dtslint fail the test if typescript does *not* generate an error on the next line (you can also check the message with `$ExpectError message`)

Your code is not failing because it simply types the `const s3` as `undefined`, since as the property is optional TS adds `undefined` to the type, which is a bit annoying 😢

Reply:
Well, despite your extensive help I still can't get it working. The code even looks identical to your playground example. Sorry for all the time this is taking!

Reply:
Ah, just remembered that CI was failing when you opened this branch due to `react-table-filter`, try rebasing/merging up to current master so you're not getting confounded by that!

Reply:
Well, it's also failing locally, so...

Reply:
Seems like it's up to date with master as well.

Reply:
What's the error you're getting?

Reply:
```
ERROR: 870:1  expect  Expected an error on this line, but found none.
ERROR: 873:1  expect  Expected an error on this line, but found none.
ERROR: 896:1  expect  Expected an error on this line, but found none.
ERROR: 899:1  expect  Expected an error on this line, but found none.
```
These are the exact lines that have the assertion `// $ExpectError`

Reply:
OK, did some playing around and it looks like a TS limitation. The reason you're seeing no errors is that you are assigning to the `never` field before reading, so TS just updates the type. For some reason this seems to cause it to merge the types so both fields are valid. 😵

At least these works, in this order:

```ts
// $ExpectError
customOrigin = request.origin.custom;

// $ExpectError
request.origin.custom.domainName = "blah";

// $ExpectError
request.origin = {
   custom: customOrigin,
   s3: s3Origin,
};
```

You will need to merge upsteam master for the CI to work though.